GTFS Trip-Modifications (#403)
google/transit: gtfs-realtime/spec/en/reference.md

| **encoded_polyline** | [string](https://protobuf.dev/programming-guides/proto2/#scalar) | Required | One | Encoded polyline representation of the shape. This polyline must contain at least two points. For more information about encoded polylines, see https://developers.google.com/maps/documentation/utilities/polylinealgorithm <br><br>**Caution:** this field is still **experimental**, and subject to change. It may be formally adopted in the future. |

## _message_ Stop

Represents a new Stop added to the feed dynamically. All fields are as described in the (CSV) GTFS specification. The location type of the new stop is `0` (routable stop). 

<br><br>**Caution:** this field is still **experimental**, and subject to change. It may be formally adopted in the future.

**Fields**

| _**Field Name**_ | _**Type**_ | _**Required**_ | _**Cardinality**_ | _**Description**_ |
|------------------|------------|----------------|-------------------|-------------------|

> | **stop_id** | [string](https://protobuf.dev/programming-guides/proto2/#scalar) | Required | One |  Identifier of the stop. Must be different than any `stop_id` defined in the (CSV) GTFS. |

Reviewer: In the proto file this field is optional. I know there was something with optional fields being recommend in protobuf but I don't remember the details.

All I know that we have a similar situation with alters where header texts are required by the spec text but not the proto and I regularly have to discuss with people if they are mandatory or not.

Reviewer 2: https://protobuf.dev/programming-guides/dos-donts/#add-required
https://news.ycombinator.com/item?id=18196288

Generally speaking, new "required" fields in protobufs are discouraged.  It's less about high-level goals of data validation and more about what happens when different data producers and consumers (and intermediates) with slightly different specs are involved + nuances of the protobuf implementation itself.

I actually thought about flagging this originally, since I was trying to think of a case where we might not specify a `stop_id` as part of `Stop` at some point down the road (maybe it's possible?).  But I would not add more required fields unnecessarily.

Author: Yes, I agree this is a bit strange but it's the way to go to avoid problem or breaking changes in the future. Part of the problem is the documentation is in the .proto too so it's used standalone. We might want to add requirement of the field in documentation in the .proto in addition to the reference.

| **start_times** | [string](https://protobuf.dev/programming-guides/proto2/#scalar) | Optional | Many | A list of start times in the real-time trip descriptor for the trip_id defined in trip_ids. Useful to target multiple departures of a trip_id in a frequency-based trip. |
| **service_dates** | [string](https://protobuf.dev/programming-guides/proto2/#scalar) | Required | Many | Dates on which the modification occurs, in the YYYYMMDD format. A trip_id will only be modified if it runs on a given service date; the trip IS NOT required to run on all of the service dates. Producers SHOULD only transmit detours occurring within the next week. The dates provided should not be used as user-facing information, if a user-facing start and end date need to be provided, they can be provided in the linked service alert with `service_alert_id` |
| **modifications** | [Modification](#message-modification) | Required | Many | A list of modifications to apply to the affected trips. |

## _message_ Modification

> A `Modification` message replaces a span of n stop times from each affected trip starting at `start_stop_selector`.

Reviewer 3: It's possible for no stop times to be replaced.

```suggestion
A `Modification` message describes changes to each affected trip starting at `start_stop_selector`.
```

Author: In that case we meant that n could be 0, but your language is clearer.

Describes the physical path that a vehicle takes when the shape is not part of the (CSV) GTFS, such as for an ad-hoc detour. Shapes belong to Trips and consist of an encoded polyline for more efficient transmission.  Shapes do not need to intercept the location of Stops exactly, but all Stops on a trip should lie within a small distance of the shape for that trip, i.e. close to straight line segments connecting the shape points

**Fields**

| _**Field Name**_ | _**Type**_ | _**Required**_ | _**Cardinality**_ | _**Description**_ |
|------------------|------------|----------------|-------------------|-------------------|
| **shape_id** | [string](https://developers.google.com/protocol-buffers/docs/proto#scalar) | Required | One |  Identifier of the shape. Must be different than any `shape_id` defined in the (CSV) GTFS. |
| **encoded_polyline** | [string](https://developers.google.com/protocol-buffers/docs/proto#scalar) | Required | One | Encoded polyline representation of the shape. This polyline must contain at least two points. For more information about encoded polylines, see https://developers.google.com/maps/documentation/utilities/polylinealgorithm|

## _message_ Stop

> Represent a new Stop added to the feed dynamically. All fields are as described in the (CSV) GTFS specification. Location type of the new stop is assumed to be `0` (routable stop). 

Author: I agree, will change.

**Fields**

| _**Field Name**_ | _**Type**_ | _**Required**_ | _**Cardinality**_ | _**Description**_ |
|------------------|------------|----------------|-------------------|-------------------|
| **shape_id** | [string](https://protobuf.dev/programming-guides/proto2/#scalar) | Required | One |  Identifier of the shape. Must be different than any `shape_id` defined in the (CSV) GTFS. |
| **encoded_polyline** | [string](https://protobuf.dev/programming-guides/proto2/#scalar) | Required | One | Encoded polyline representation of the shape. This polyline must contain at least two points. For more information about encoded polylines, see https://developers.google.com/maps/documentation/utilities/polylinealgorithm|

## _message_ Stop

Represents a new Stop added to the feed dynamically. All fields are as described in the (CSV) GTFS specification. The location type of the new stop is `0` (routable stop). 

**Fields**

| _**Field Name**_ | _**Type**_ | _**Required**_ | _**Cardinality**_ | _**Description**_ |
|------------------|------------|----------------|-------------------|-------------------|

> | **stop_id** | [string](https://protobuf.dev/programming-guides/proto2/#scalar) | Required | One |  Identifier of the shape. Must be different than any `stop_id` defined in the (CSV) GTFS. |

Reviewer 2: nit: Identifier of the shape => Identifier of the stop.

| **wheelchair_boarding** | [WheelchairBoarding](#enum-wheelchairboarding) | Optional | One |  See definition of [stops.wheelchair_boarding](https://github.com/google/transit/blob/master/gtfs/spec/en/reference.md#stopstxt) in (CSV) GTFS. |
| **level_id** | [string](https://protobuf.dev/programming-guides/proto2/#scalar) | Optional | One |  See definition of [stops.level_id](https://github.com/google/transit/blob/master/gtfs/spec/en/reference.md#stopstxt) in (CSV) GTFS. |
| **platform_code** | [TranslatedString](#message-translatedstring) | Optional | One |  See definition of [stops.platform_code](https://github.com/google/transit/blob/master/gtfs/spec/en/reference.md#stopstxt) in (CSV) GTFS. |

## _enum_ WheelchairBoarding

**Values**

| _**Value**_ | _**Comment**_ |
|-------------|---------------|
| **UNKNOWN** | No accessibility information for the stop. |
| **AVAILABLE** |  Some vehicles at this stop can be boarded by a rider in a wheelchair. |
| **NOT_AVAILABLE** | Wheelchair boarding is not possible at this stop. |

## _message_ TripModifications

> A `TripModifications` message identifies a list of similar trips which are all affected by a particular modification, such as a detour.

Reviewer 4: ```suggestion
A `TripModifications` message identifies a list of similar trips which are all affected by particular modifications, such as a detour.
```
